File: spec/system/debugging_case_validation_spec.rb
# frozen_string_literal: true

require "rails_helper"

RSpec.describe "Debugging case validation", type: :system do
  before do
    user = create(:user, email: "validation-#{SecureRandom.hex(4)}@example.com")
    sign_in_via_browser(email: user.email)
  end

  it "re-renders the new case form with validation errors when no log sources are provided" do
    visit new_debugging_case_path

    fill_in "Title", with: "Missing sources case"
    click_button "Create debugging case"

    expect(page).to have_current_path(debugging_cases_path)
    expect(page).to have_css("h1", text: "New debugging case")
    expect(page).to have_text("prohibited this case from being saved")
    expect(page).to have_text("must include at least one non-blank log source")
    expect(page).to have_field("Title", with: "Missing sources case")
    expect(page).to have_css(".form-field--invalid")
  end

  it "preserves metadata fields when validation fails" do
    visit new_debugging_case_path

    fill_in "Title", with: "Metadata preserve case"
    fill_in "Description", with: "Reporter notes"
    fill_in "Customer reference", with: "Ticket #999"
    fill_in "Environment", with: "staging"
    click_button "Create debugging case"

    expect(page).to have_css("h1", text: "New debugging case")
    expect(page).to have_field("Title", with: "Metadata preserve case")
    expect(page).to have_field("Description", with: "Reporter notes")
    expect(page).to have_field("Customer reference", with: "Ticket #999")
    expect(page).to have_field("Environment", with: "staging")
  end

  it "preserves source metadata and highlights invalid slots without re-rendering paste" do
    visit new_debugging_case_path

    fill_in "Title", with: "Invalid source type case"
    within(find("fieldset", text: "Log source 1")) do
      fill_in "Name (optional)", with: "Rails"
      fill_in "Pasted content", with: "secret-token-#{SecureRandom.hex(4)}"
    end
    click_button "Create debugging case"

    expect(page).to have_css("h1", text: "New debugging case")
    expect(page).to have_css("fieldset.fieldset--invalid", text: "Log source 1")
    expect(page).to have_field("Name (optional)", with: "Rails")
    expect(page).to have_field("Pasted content", with: "")
    expect(page).to have_text("Pasted log content was cleared for security")
  end
end

File: e2e/debugging-case-validation.spec.ts
import path from "path";
import { test, expect } from "@playwright/test";
import {
  createAuthenticatedContext,
  saveAuthStorageState,
  signUp,
  uniqueEmail,
} from "./helpers";

const authFile = path.join(__dirname, ".auth", "validation-user.json");

test.describe("Debugging case validation", () => {
  test.beforeAll(async ({ browser }) => {
    const context = await browser.newContext();
    const page = await context.newPage();
    const email = uniqueEmail("pw-validation");

    await signUp(page, email);
    await saveAuthStorageState(page, authFile);
    await context.close();
  });

  test("re-renders the new case form with validation errors when no log sources are provided", async ({
    browser,
  }) => {
    const context = await createAuthenticatedContext(browser, authFile);
    const page = await context.newPage();

    await page.goto("/debugging_cases/new");
    await page.getByLabel("Title").fill("Missing sources case");
    await page.getByRole("button", { name: "Create debugging case" }).click();

    await expect(page).toHaveURL(/\/debugging_cases$/);
    await expect(page.getByRole("heading", { name: "New debugging case" })).toBeVisible();
    await expect(page.getByText("prohibited this case from being saved")).toBeVisible();
    await expect(page.locator("#error_explanation")).toContainText(
      "must include at least one non-blank log source",
    );
    await expect(page.getByLabel("Title")).toHaveValue("Missing sources case");
    await expect(page.locator(".form-field--invalid").first()).toBeVisible();

    await context.close();
  });

  test("preserves metadata fields when validation fails", async ({ browser }) => {
    const context = await createAuthenticatedContext(browser, authFile);
    const page = await context.newPage();

    await page.goto("/debugging_cases/new");
    await page.getByLabel("Title").fill("Metadata preserve case");
    await page.getByLabel("Description").fill("Reporter notes");
    await page.getByLabel("Customer reference").fill("Ticket #999");
    await page.getByLabel("Environment").fill("staging");
    await page.getByRole("button", { name: "Create debugging case" }).click();

    await expect(page.getByRole("heading", { name: "New debugging case" })).toBeVisible();
    await expect(page.getByLabel("Title")).toHaveValue("Metadata preserve case");
    await expect(page.getByLabel("Description")).toHaveValue("Reporter notes");
    await expect(page.getByLabel("Customer reference")).toHaveValue("Ticket #999");
    await expect(page.getByLabel("Environment")).toHaveValue("staging");

    await context.close();
  });

  test("preserves source metadata and shows paste-cleared hint on invalid source type", async ({
    browser,
  }) => {
    const context = await createAuthenticatedContext(browser, authFile);
    const page = await context.newPage();
    const sourceName = "Rails E2E";

    await page.goto("/debugging_cases/new");
    await page.getByLabel("Title").fill("Invalid source type E2E case");
    await page.getByLabel("Name (optional)", { exact: true }).first().fill(sourceName);
    await page.getByLabel("Pasted content").first().fill("e2e-paste-should-not-rerender");
    await page.getByRole("button", { name: "Create debugging case" }).click();

    await expect(page.getByRole("heading", { name: "New debugging case" })).toBeVisible();
    await expect(page.locator("fieldset.fieldset--invalid").first()).toBeVisible();
    await expect(page.getByLabel("Name (optional)", { exact: true }).first()).toHaveValue(sourceName);
    await expect(page.getByLabel("Pasted content").first()).toHaveValue("");
    await expect(page.getByText("Pasted log content was cleared for security")).toBeVisible();

    await context.close();
  });
});
